tracer/src/Datadog.Trace/FeatureFlags/FeatureFlagsEvaluator.cs

```diff
@@ -0,0 +1,672 @@
+
+        public Evaluation Evaluate(string flagKey, ValueType resultType, object? defaultValue, IEvaluationContext? context)
+        {
+            try
+            {
+                var config = _config;
+                if (config == null)
+                {
+                    return new Evaluation(
+                        flagKey,
+                        defaultValue,
+                        EvaluationReason.Error,
+                        error: "PROVIDER_NOT_READY",
+                        metadata: new Dictionary<string, string>
+                        {
+                            ["errorCode"] = "PROVIDER_NOT_READY"
+                        });
+                }
+
+                if (string.IsNullOrEmpty(context?.TargetingKey))
+                {
+                    return new Evaluation(
+                        flagKey,
+                        defaultValue,
+                        EvaluationReason.Error,
+                        error: "TARGETING_KEY_MISSING",
+                        metadata: new Dictionary<string, string>
+                        {
+                            ["errorCode"] = "TARGETING_KEY_MISSING"
+                        });
+                }
+
+                if (config.Flags is null || !config.Flags.TryGetValue(flagKey, out var flag) || flag is null)
+                {
+                    return new Evaluation(
+                        flagKey,
+                        defaultValue,
+                        EvaluationReason.Error,
+                        error: "FLAG_NOT_FOUND",
+                        metadata: new Dictionary<string, string>
+                        {
+                            ["errorCode"] = "FLAG_NOT_FOUND"
+                        });
+                }
+
+                if (!flag.Enabled is true)
+                {
+                    return new Evaluation(
+                        flagKey,
+                        defaultValue,
+                        EvaluationReason.Disabled);
+                }
+
+                if (flag.VariationType != resultType)
+                {
+                    return new Evaluation(
+                        flagKey,
+                        defaultValue,
+                        EvaluationReason.Error,
+                        error: "TYPE_MISMATCH",
+                        metadata: new Dictionary<string, string>
+                        {
+                            ["errorCode"] = "TYPE_MISMATCH"
+                        });
+                }
+
+                if (flag.Allocations is { Count: 0 })
+                {
+                    return new Evaluation(
+                        flagKey,
+                        defaultValue,
+                        EvaluationReason.Default);
+                }
+
+                var now = DateTime.UtcNow;
+                var targetingKey = context?.TargetingKey ?? string.Empty;
+
+                foreach (var allocation in flag.Allocations!)
```

Comment:
`flag.Allocations` could be null here, how should we handle that? We could change the previous check to this instead

```csharp
                if (flag.Allocations is not { Count: > 0 })
```

which would fix this:


```suggestion
                foreach (var allocation in flag.Allocations)
```